Add Hibernate Search management endpoint

marko-bekhta opened the pull request:
Fixes #34553

Opening this as a draft so we can discuss it. A couple of questions from my side:

- Is it ok for management endpoints to be more complex than simple GET, if so
  - Do we want the mass-indexer configuration to be available at all? For now, I've added it as config properties, but I was wondering if we maybe should move it to the body of a POST request or maybe allow both options?
  - Do we want to be more flexible and allow indexing not all but a subset of entities? There's a filter in a query params for now that can do it, but again, maybe it would be better to move it to the request body?
  - Same about tenants, with an improvement we've added to Search mass-indexer should pick all tenants automatically, but maybe we want to have a  filter for tenants too?
- How about more granular operations like reindexing a specific entity (by ID)? We could do it through an indexing plan...

I remember implementing similar logic in the webapps quite often, and if these endpoints would be available to an admin in prod mode, it could save people some time 😃

marko-bekhta committed:
Add Hibernate Search management endpoint

- to reindex either all or subset of indexed entities

yrodiere commented on extensions/hibernate-search-orm-elasticsearch/runtime/src/main/java/io/quarkus/hibernate/search/orm/elasticsearch/runtime/management/HibernateSearchManagementConfig.java:
```diff
@@ -0,0 +1,101 @@
+        /**
+         * Defines whether the mass indexer should wait until indexing is completed before returning the response to the user.
+         */
+        @WithDefault("false")
+        Boolean blocking();
```
Probably should be a query parameter.

yrodiere commented on extensions/hibernate-search-orm-elasticsearch/runtime/src/main/java/io/quarkus/hibernate/search/orm/elasticsearch/runtime/management/HibernateSearchManagementConfig.java:
```diff
@@ -0,0 +1,101 @@
+
+        /**
+         * Sets the batch size used to load the root entities.
+         */
+        Optional<Integer> batchSizeToLoadObjects();
+
+        /**
+         * Sets the cache interaction mode for the data loading tasks.
+         */
+        Optional<CacheMode> cacheMode();
+
+        /**
+         * If each index is merged into a single segment after indexing.
+         */
+        Optional<Boolean> mergeSegmentsOnFinish();
+
+        /**
+         * If each index is merged into a single segment after the initial index purge, just before indexing.
+         */
+        Optional<Boolean> mergeSegmentsAfterPurge();
+
+        /**
+         * If the indexes and their schema (if they exist) should be dropped and re-created before indexing.
+         */
+        Optional<Boolean> dropAndCreateSchemaOnStart();
+
+        /**
+         * If all entities are removed from the indexes before indexing.
+         */
+        Optional<Boolean> purgeAllOnStart();
+
+        /**
+         * Specifies the fetch size to be used when loading primary keys if objects to be indexed.
+         */
+        Optional<Integer> idFetchSize();
+
+        /**
+         * Specifies the timeout of transactions for loading ids and entities to be re-indexed.
+         */
+        Optional<Integer> transactionTimeout();
```
Probably should be passed through a POST payload (JSON?)

yrodiere commented on integration-tests/hibernate-search-orm-elasticsearch/src/main/resources/application.properties:
```diff
@@ -20,4 +20,13 @@ quarkus.hibernate-search-orm.indexing.plan.synchronization.strategy=sync
+# now enable the Hibernate Search management itself:
+quarkus.hibernate-search.management.enabled=true
+# and let's use some "custom" base url to see that the config property takes effect:
+quarkus.hibernate-search.management.root-path=reeeeeindex
```
Could you test the default route path separately, please?

yrodiere commented on integration-tests/hibernate-search-orm-elasticsearch/src/test/java/io/quarkus/it/hibernate/search/orm/elasticsearch/management/HibernateSearchManagementTest.java:
```diff
@@ -0,0 +1,48 @@
+import io.restassured.RestAssured;
+
+@QuarkusTest
+@DisabledOnOs(OS.WINDOWS)
```
Might I ask why? :)

marko-bekhta replied:
I've noticed that the tests for dev services were also disabled like that for Windows, so I've assumed that "fewer tests executed on that env is better" was the goal here 😄...

yrodiere replied:
> I've noticed that the tests for dev services were also disabled like that for Windows, so I've assumed that "fewer tests executed on that env is better" was the goal here 😄...

Hey @gsmet, could you confirm ^, please?

marko-bekhta replied:
To clarify, I was talking about the tests in `io.quarkus.it.hibernate.search.orm.elasticsearch.devservices` 😃

yrodiere replied:
Ok, in that case this probably doesn't make sense here.

yrodiere reviewed (changes requested):
Thanks! Comments below.

> Is it ok for management endpoints to be more complex than simple GET

Well for a start, in this case it most definitely should be a POST, because GET semantics are not a good match :)

And I guess you should be able to accept a payload, yes.

> Do we want the mass-indexer configuration to be available at all

It's certainly useful, I'm just not sure we want to maintain this.

But if you've already started working on this... let's say "yes, but be sure to mark this as preview technology in the documentation"? I don't see any documentation, btw :)

> Do we want to be more flexible and allow indexing not all but a subset of entities? There's a filter in a query params for now that can do it, but again, maybe it would be better to move it to the request body?

Makes sense. I'd suggest relying on entity type names: `searchMapping.scope( Object.class, Arrays.asList( "entityName1", "entityName2" ) )`

> Same about tenants, with an improvement we've added to Search mass-indexer should pick all tenants automatically, but maybe we want to have a filter for tenants too?

Makes sense.

> How about more granular operations like reindexing a specific entity (by ID)? We could do it through an indexing plan...

Looks interesting, and indeed I've used such things myself, but I'd suggest creating a separate ticket as we need to refine the requirements: should this trigger reindexing of containing classes (in which case `IndexingPlan` makes sense)? Or just the entity we're being pointed to (in which case `MassIndexer` with a [condition](https://docs.jboss.org/hibernate/stable/search/reference/en-US/html_single/#indexing-massindexer-conditional) is more appropriate).

> I remember implementing similar logic in the webapps quite often

I'd be interested in discussing those webapps (offline); I thought you hadn't used Hibernate Search much before joining the team :)

marko-bekhta commented:
Hey Yoann 😃 
Here's the next iteration on this one. 
- I've added support for non-default PUs through an optional query param.
- There's some documentation now.
- Moved the config of the mass indexer to the body of the request.
- Same for filters on tenants or types to reindex -- they are now in the request body too.

I've also noticed that there seems to be a problem with Flyway... `quarkus.flyway.clean-at-start` cleans the records (including the ones in the `flyway_schema_history`, but it keeps the tables. This means if the tests are executed against the same container, they try to run the migration, but that fails because all the DB objects are still there... so I've added a couple of drop statements to the script to address the issue for the tests. But overall this looks like a bug to me...

yrodiere commented on extensions/hibernate-search-orm-elasticsearch/runtime/src/main/java/io/quarkus/hibernate/search/orm/elasticsearch/runtime/management/HibernateSearchPostRequestProcessor.java:
```diff
@@ -0,0 +1,214 @@
+                    massIndexer, config.getJsonObject("massIndexer"));
+
+            if (isBlockingIndexerRequested(ctx.request())) {
+                try {
+                    massIndexer.startAndWait();
+                } catch (InterruptedException e) {
+                    Thread.currentThread().interrupt();
+                    response(ctx, "error");
+                }
+                response(ctx, "done");
+            } else {
+                massIndexer.start();
+                response(ctx, "started");
```
I'm pretty sure blocking here like you do in `massIndexer.startAndWait()` is a big no-no.

When you're asked to block, couldn't you rather pass the future returned by `massIndexer.start()` to the response context, and let Vertx handle the waiting itself?

Also, you might want to have a look at what I did in this demo (I took inspiration from something else in Quarkus, don't remember what): https://github.com/hibernate/hibernate-demos/pull/74/files#diff-e2ecf487e7108639ecb9c840b4fea7cca821fd31f63c48cb91ac85f608bb3765
Basically it implements chunked transfer: when mass indexing starts it sends one chunk with a JSON object aknowledging the request, and when it stops it sends another chunk with a JSON object describing the result.

I'm not saying it absolutely needs to be in this PR, but it might be a good idea to think about how we'd add it, and what we would need to change. I'm thinking that "blocking=true" might not to be expanded to something like `wait_for=started` (response returned when mass indexing started) / `wait_for=progress` (response returns chunked progress) / `wait_for=finished` (response returned when mass indexing finished).
So maybe use `wait_for=started`/`wait_for=finished` for now, so that we can easily add the third option later? Or another name that also fits, I'm just suggesting `wait_for` because Elasticsearch has [something similar](https://www.elastic.co/guide/en/elasticsearch/reference/current/docs-refresh.html).

Two culprits to this "progress" thing though:

* Maybe there's a standard in HTTP to specify whether the request should block or not, should use chunked encoding or not? Some header maybe? I'm really not sure.
* Maybe there's a standard schema the JSON objects for the progress could follow.

yrodiere commented on docs/src/main/asciidoc/hibernate-search-orm-elasticsearch.adoc:
```diff
@@ -1015,6 +1015,88 @@ You can enable AWS request signing in Hibernate Search by adding a dedicated ext
+[[management]]
+== Management endpoint
+
+:extension-status: preview
+include::{includes}/extension-status.adoc[]
```
I'm not sure it makes sense to do this at the section level; it's not the whole extension we're talking about, but just one feature.

Maybe do something similar to what we did for outbox-polling coordination instead? See https://quarkus.io/guides/hibernate-search-orm-elasticsearch#coordination

yrodiere commented on docs/src/main/asciidoc/hibernate-search-orm-elasticsearch.adoc:
```diff
@@ -1015,6 +1015,88 @@ You can enable AWS request signing in Hibernate Search by adding a dedicated ext
+:extension-status: preview
+include::{includes}/extension-status.adoc[]
+
+Hibernate Search extension has a capability to reindex the data through the xref:./management-interface-reference.adoc[management interface].
```
```suggestion
The Hibernate Search extension provides an HTTP endpoint to reindex your data through the xref:./management-interface-reference.adoc[management interface].
```

yrodiere commented on docs/src/main/asciidoc/hibernate-search-orm-elasticsearch.adoc:
```diff
@@ -1015,6 +1015,88 @@ You can enable AWS request signing in Hibernate Search by adding a dedicated ext
+<2> Enable Hibernate Search specific management endpoints.
+
+Once the management is enabled, data can be re-indexed via `/q/reindex`, where `/q` is the default management root path.
+Reindex path (`/reindex`) can be changed via configuration property as shown below.
```
```suggestion
The reindexing path (`/reindex`) can be changed via configuration property as shown below.
```

yrodiere commented on docs/src/main/asciidoc/hibernate-search-orm-elasticsearch.adoc:
```diff
@@ -1015,6 +1015,88 @@ You can enable AWS request signing in Hibernate Search by adding a dedicated ext
+
+[source,properties]
+----
+quarkus.hibernate-search.management.root-path=custom-root-path <1>
```
Feels weird that `/reindex` is called the "root" path... it would mean that if we ever add another endpoint to the management interface, it can only be named `/reindex/something`?

Wouldn't it make sense for the default to be something like `/q/hibernate-search/reindex` and for the "root path" to only change "hibernate-search" (e.g. it'd become `/q/custom-root-path/reindex`)?

yrodiere commented on docs/src/main/asciidoc/hibernate-search-orm-elasticsearch.adoc:
```diff
@@ -1015,6 +1015,88 @@ You can enable AWS request signing in Hibernate Search by adding a dedicated ext
+If the default management root path is used then the reindex path becomes `/q/custom-root-path`.
+
+This endpoint accepts `POST` requests with `application/json` content type only.
+All indexed entities will be re-indexed if an empty request body is submitted.
```
Maybe add a simple example here? Just so that people don't think they need to put all the complicated configuration below.

yrodiere commented on docs/src/main/asciidoc/hibernate-search-orm-elasticsearch.adoc:
```diff
@@ -1015,6 +1015,88 @@ You can enable AWS request signing in Hibernate Search by adding a dedicated ext
+  }
+}
+----
+<1> Filter object that allows to limit the amount of indexed entities.
```
```suggestion
<1> Filter object that allows to limit the scope of reindexing.
```

yrodiere commented on docs/src/main/asciidoc/hibernate-search-orm-elasticsearch.adoc:
```diff
@@ -1015,6 +1015,88 @@ You can enable AWS request signing in Hibernate Search by adding a dedicated ext
+}
+----
+<1> Filter object that allows to limit the amount of indexed entities.
+<2> An array of entity names that should be re-indexed, if empty -- all entities will be re-indexed.
```
```suggestion
<2> An array of entity names that should be re-indexed. If unspecified or empty, all entity types will be re-indexed.
```

yrodiere commented on docs/src/main/asciidoc/hibernate-search-orm-elasticsearch.adoc:
```diff
@@ -1015,6 +1015,88 @@ You can enable AWS request signing in Hibernate Search by adding a dedicated ext
+----
+<1> Filter object that allows to limit the amount of indexed entities.
+<2> An array of entity names that should be re-indexed, if empty -- all entities will be re-indexed.
+<3> An array of tenant ids. In case of multi tenancy, re-indexing can be limited to a subset of available tenants.
+If empty data for all tenants will be re-indexed.
```
```suggestion
<3> An array of tenant ids, in case of multi-tenancy. If unspecified or empty, all tenants will be re-indexed.
```

yrodiere commented on docs/src/main/asciidoc/hibernate-search-orm-elasticsearch.adoc:
```diff
@@ -1015,6 +1015,88 @@ You can enable AWS request signing in Hibernate Search by adding a dedicated ext
+<6> Sets the number of threads to be used to load the root entities.
+<7> Sets the batch size used to load the root entities.
+<8> Sets the cache interaction mode for the data loading tasks.
+<9> If each index is merged into a single segment after indexing.
+<10> If each index is merged into a single segment after the initial index purge, just before indexing.
+<11> If the indexes and their schema (if they exist) should be dropped and re-created before indexing.
+<12> If all entities are removed from the indexes before indexing.
```
```suggestion
<9> Whether each index is merged into a single segment after indexing.
<10> Whether each index is merged into a single segment after the initial index purge, just before indexing.
<11> Whether the indexes and their schema (if they exist) should be dropped and re-created before indexing.
<12> Whether all entities are removed from the indexes before indexing.
```

yrodiere commented on docs/src/main/asciidoc/hibernate-search-orm-elasticsearch.adoc:
```diff
@@ -1015,6 +1015,88 @@ You can enable AWS request signing in Hibernate Search by adding a dedicated ext
+For more detailed information on mass indexer configuration see the
+link:{hibernate-search-docs-url}#indexing-massindexer-parameters[corresponding section of the Hibernate Search reference documentation].
+
+Submitting the reindexing request will trigger the indexing in the background.
```
```suggestion
Submitting the reindexing request will trigger indexing in the background. Mass indexing progress will appear in the application logs.
```

yrodiere commented on docs/src/main/asciidoc/hibernate-search-orm-elasticsearch.adoc:
```diff
@@ -1015,6 +1015,88 @@ You can enable AWS request signing in Hibernate Search by adding a dedicated ext
+
+Submitting the reindexing request will trigger the indexing in the background.
+For testing purposes, it might be useful to know when the indexing finished. Adding `blocking=true` query parameter to the URL
+will make mass indexer run in the blocking mode and response will not be returned until mass indexer finishes the process.
```
```suggestion
will make mass indexer run in the blocking mode and the response will only be returned when mass indexer finishes the process.
```

yrodiere commented on docs/src/main/asciidoc/hibernate-search-orm-elasticsearch.adoc:
```diff
@@ -1015,6 +1015,88 @@ You can enable AWS request signing in Hibernate Search by adding a dedicated ext
+For testing purposes, it might be useful to know when the indexing finished. Adding `blocking=true` query parameter to the URL
+will make mass indexer run in the blocking mode and response will not be returned until mass indexer finishes the process.
+
+When working with multiple persistence units the name of the persistence unit to reindex can be supplied through the
```
```suggestion
When working with multiple persistence units, the name of the persistence unit to reindex can be supplied through the
```

yrodiere reviewed (commented):
Thanks. A few comments below.

> I've also noticed that there seems to be a problem with Flyway... quarkus.flyway.clean-at-start cleans the records (including the ones in the flyway_schema_history, but it keeps the tables. This means if the tests are executed against the same container, they try to run the migration, but that fails because all the DB objects are still there... so I've added a couple of drop statements to the script to address the issue for the tests. But overall this looks like a bug to me...

No idea what is going on. Maybe open a bug report?

yrodiere commented on extensions/hibernate-search-orm-elasticsearch/runtime/src/main/java/io/quarkus/hibernate/search/orm/elasticsearch/runtime/management/HibernateSearchPostRequestProcessor.java:
```diff
@@ -0,0 +1,220 @@
+            CompletionStage<?> massIndexerFuture = massIndexer.start();
+
+            if (WaitFor.STARTED.equals(getWaitForParameter(ctx.request()))) {
+                ctx.response().end(message(202, "Reindex started"));
+            } else {
+                ctx.response()
+                        .setChunked(true)
+                        .write(message(202, "Reindex started"),
+                                ignored -> massIndexerFuture.whenComplete((ignored2, throwable) -> {
+                                    if (throwable == null) {
+                                        ctx.response().end(message(200, "Reindex succeeded"));
+                                    } else {
+                                        ctx.response().end(message(
+                                                500,
+                                                "Reindex failed:\n" + Arrays.stream(throwable.getStackTrace())
+                                                        .map(Object::toString)
+                                                        .collect(Collectors.joining("\n"))));
+                                    }
```
```suggestion
                ctx.response().end(message(202, "Reindexing started"));
            } else {
                ctx.response()
                        .setChunked(true)
                        .write(message(202, "Reindexing started"),
                                ignored -> massIndexerFuture.whenComplete((ignored2, throwable) -> {
                                    if (throwable == null) {
                                        ctx.response().end(message(200, "Reindexing succeeded"));
                                    } else {
                                        ctx.response().end(message(
                                                500,
                                                "Reindexing failed:\n" + Arrays.stream(throwable.getStackTrace())
                                                        .map(Object::toString)
                                                        .collect(Collectors.joining("\n"))));
                                    }
```

yrodiere commented on extensions/hibernate-search-orm-elasticsearch/deployment/src/main/java/io/quarkus/hibernate/search/orm/elasticsearch/deployment/HibernateSearchElasticsearchProcessor.java:
```diff
@@ -435,4 +438,22 @@ void devServicesDropAndCreateAndDropByDefault(
+                .routeConfigKey("quarkus.hibernate-search-orm.management.root-path")
+                .handler(recorder.managementHandler())
+                .displayOnNotFoundPage()
+                .blockingRoute()
```
It's not blocking, is it?

```suggestion
```

yrodiere commented:
Thanks @marko-bekhta , LGTM. I'll mark this as ready for review, since... it's already been reviewed :) Let's see what CI has to say about it.

yrodiere commented:
@marko-bekhta Tests seem to fail, see ^

marko-bekhta commented:
yeah 😔 
```
java.lang.NoClassDefFoundError: io/quarkus/vertx/http/deployment/RouteBuildItem
	at java.base/java.lang.Class.forName0(Native Method)
	at java.base/java.lang.Class.forName(Class.java:496)
```
this is coming from the 
```xml
        <dependency>
            <groupId>io.quarkus</groupId>
            <artifactId>quarkus-vertx-http-deployment</artifactId>
            <optional>true</optional>
        </dependency>
```
being an optional dependency... But I'm not sure whether it can be non-optional in `quarkus-hibernate-search-orm-elasticsearch-deployment`. Should a management part be an additional artifact? 

The other failure is from the label change... so that's easier to fix 😃

yrodiere commented:
> java.lang.NoClassDefFoundError: io/quarkus/vertx/http/deployment/RouteBuildItem
> 	at java.base/java.lang.Class.forName0(Native Method)
> 	at java.base/java.lang.Class.forName(Class.java:496)
> ```
> 
> this is coming from the
> 
> ```
>         <dependency>
>             <groupId>io.quarkus</groupId>
>             <artifactId>quarkus-vertx-http-deployment</artifactId>
>             <optional>true</optional>
>         </dependency>
> ```
> 
> being an optional dependency...

Ok, makes sense.

>  But I'm not sure whether it can be non-optional in `quarkus-hibernate-search-orm-elasticsearch-deployment`.

It is non-optional for Micrometer at least:

https://github.com/quarkusio/quarkus/blob/351195aab2c19fcbe441bc975ba294c90da21ed3/extensions/micrometer/deployment/pom.xml#L35-L38

If we're talking about the deployment artifact only, I think the non-optional dependency is fine? It won't be present at runtime, only at build time.

TBH I'm not sure why the dependency was optional in the first place.

> Should a management part be an additional artifact?

No I don't think we need an additional artifact here.

marko-bekhta commented:
Hey hey @cescoffier 

We are trying to add some optional management endpoints to the Hibernate Search Extension. Having 

```xml
<dependency>
    <groupId>io.quarkus</groupId>
    <artifactId>quarkus-vertx-http-deployment</artifactId>
    <optional>true</optional>
</dependency>
```
as a dependency in build (and `quarkus-vertx-http` in runtime) module of the extension works for as long as `quarkus-vertx-http` is also available at runtime. Otherwise the build fails with 
```
java.lang.NoClassDefFoundError: io/quarkus/vertx/http/deployment/RouteBuildItem
	at java.base/java.lang.Class.forName0(Native Method)
	at java.base/java.lang.Class.forName(Class.java:496)
```

Making the build dependency required while leaving the runtime dependency as optional doesn't work since it leads to:

```
[ERROR] Failed to execute goal io.quarkus:quarkus-extension-maven-plugin:999-SNAPSHOT:extension-descriptor (generate-extension-descriptor) on project quarkus-hibernate-search-orm-elasticsearch: The deployment artifact io.quarkus:quarkus-hibernate-search-orm-elasticsearch-deployment::jar depends on the following Quarkus extension deployment artifacts whose corresponding runtime artifacts were not found among the dependencies of io.quarkus:quarkus-hibernate-search-orm-elasticsearch:jar:999-SNAPSHOT: io.quarkus:quarkus-netty-deployment::jar io.quarkus:quarkus-vertx-deployment::jar io.quarkus:quarkus-vertx-http-deployment::jar io.quarkus:quarkus-virtual-threads-deployment::jar
```

Yoann suggested to ping you for an idea/example on how we could make these management endpoints optional and only available if the user has `quarkus-vertx-http` without making it a mandatory dependency of the Hibernate Search extension. Any advice or pointers would be much appreciated. Thanks!

cescoffier commented:
I need to think about it. We can move it to an SPI module, but there is more than this build item. How it's managed (main router, application router, management endpoint) is particularly tricky, and I think it will drag a bit of this logic in the SPI (which is not something great, as it's an internal detail). 

Don't we have other cases where the routes are optional? In micrometer, we use a lot of optional extensions (so, I would look into a pattern there).

yrodiere commented:
> Don't we have other cases where the routes are optional? In micrometer, we use a lot of optional extensions (so, I would look into a pattern there).

From what I can see Micrometer depends on "optional" extensions at runtime only (no build items required from those optional extensions). In our case we need to depend on `quarkus-vertx-http` at build time, too...

@marko-bekhta did you try a non-optional dependency at build time (`deployment` module) + an optional dependency at runtime (`runtime` module)?

marko-bekhta commented:
> did you try a non-optional dependency at build time (deployment module) + an optional dependency at runtime (runtime module)?

Yeah 😔, that's when that maven build failure happens because of the check:

```
[INFO] 12 goals, 12 executed
[ERROR] Failed to execute goal io.quarkus:quarkus-extension-maven-plugin:999-SNAPSHOT:extension-descriptor (generate-extension-descriptor) on project quarkus-hibernate-search-orm-elasticsearch: The deployment artifact io.quarkus:quarkus-hibernate-search-orm-elasticsearch-deployment::jar depends on the following Quarkus extension deployment artifacts whose corresponding runtime artifacts were not found among the dependencies of io.quarkus:quarkus-hibernate-search-orm-elasticsearch:jar:999-SNAPSHOT: io.quarkus:quarkus-netty-deployment::jar io.quarkus:quarkus-vertx-deployment::jar io.quarkus:quarkus-vertx-http-deployment::jar io.quarkus:quarkus-virtual-threads-deployment::jar -> 
```

yrodiere commented:
Right. I guess that makes sense since `quarkus-vertx-http-deployment` could trigger generation of bytecode that depends on its runtime, so we can't allow the runtime to be optional.

Hence the suggestion to introduce an SPI... if that SPI only declares dependencies + build items, and does not trigger any bytecode generation by itself, then it becomes fine to depend on that SPI in `-deployment`, and on the runtime in our own runtime (with `<optional>true</optional>`).

yrodiere commented:
Hey @cescoffier, any news on that `quarkus-vertx-http-deployment-spi` that would enable declaring management interfaces without pulling in vertx-http into all extensions?
Should we create an issue about it?

yrodiere commented:
> Please create an issue, I didn't look at it yet.

Done: https://github.com/quarkusio/quarkus/issues/37144

yrodiere commented:
Yay, it works :)

Let's wait for https://github.com/quarkusio/quarkus/pull/38143 to be merged first though :)

gsmet commented:
I rebased it on top of Clément's PR.

marko-bekhta commented:
Thanks! I was about to go fetch a laptop to rebase it 🙈 😃
